### PR TITLE
CDAP-2490 Check if invalid Run Records for Spark and MapReduce part of run from Workflow child programs

### PR DESCRIPTION
**Pull request by henry-cask**
In 3.0 we have a scheduler thread to check and update Run Record store for those with RUNNING state but does not have the real program running via check to RuntimeService.

However, for Run Records for Workflow child programs, the runtime service never return runtime info since those runs are run from Workflow driver. Which means that there is no Twill application for the internal child programs run records to check against.

The effect is that the UI will display children program runs for a workflow as FAILED even though it may be COMPLETED or still RUNNING.

This PR contains fix to update the RunRecord checker to account for this situation:
-) Refactor the code to reuse the logic to get the right Program id for run records with RUNNING state but not actually running.
-) Check the RunRecord properties for "workflowrunid" property which contains Workflow parent run id then check the status of that run. If it is in RUNNING status but no actual program still running then it is in invalid state to update children run records status accordingly.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java [hunk 1]
```diff
@@ -242,107 +239,153 @@ public boolean apply(@Nullable RunRecord input) {
     });
 
     if (!invalidRunRecords.isEmpty()) {
-      LOG.debug("Found {} RunRecords with RUNNING status but the program not actually running.",
-                invalidRunRecords.size());
+      LOG.debug("Found {} RunRecords with RUNNING status but the program not actually running for type {}",
```
Comment (sagarkapare) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: should it be `warn` by default?

**Comment by sagarkapare** [comment]
Minor comment, otherwise LGTM.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java (hunk 1 above)
Comment (henry-cask) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java]: Sounds good, will update to warn level.

**Comment by henry-cask** [comment]
Build:
http://builds.cask.co/browse/CDAP-DUT1819-2

**Comment by henry-cask** [comment]
Thanks @sagarkapare, build successful. Merging it now ...

**Pull request merged**